pkg/oc/experimental/dockergc/client.go

```diff
@@ -0,0 +1,90 @@
+package dockergc
+
+import (
+	"context"
+	"time"
+
+	dockerapi "github.com/docker/engine-api/client"
+	dockertypes "github.com/docker/engine-api/types"
+)
+
+type dockerClient struct {
+	// timeout is the timeout of short running docker operations.
+	timeout time.Duration
+	// docker API client
+	client *dockerapi.Client
+}
+
+func newDockerClient(timeout time.Duration) (*dockerClient, error) {
+	client, err := dockerapi.NewEnvClient()
+	if err != nil {
+		return nil, err
+	}
+	return &dockerClient{
+		client:  client,
+		timeout: timeout,
+	}, nil
+}
+
+func (c *dockerClient) getTimeoutContext() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), c.timeout)
+}
+
+func (c *dockerClient) Info() (*dockertypes.Info, error) {
+	ctx, cancel := c.getTimeoutContext()
+	defer cancel()
+	info, err := c.client.Info(ctx)
+	if ctx.Err() == context.DeadlineExceeded {
+		return nil, ctx.Err()
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &info, nil
+}
+
+func (c *dockerClient) ContainerList(options dockertypes.ContainerListOptions) ([]dockertypes.Container, error) {
+	ctx, cancel := c.getTimeoutContext()
+	defer cancel()
+	containers, err := c.client.ContainerList(ctx, options)
+	if ctx.Err() == context.DeadlineExceeded {
```

Comment:
Why do you need this additional check?

Reply:
There are two places an error can happen: the ctx can timeout or the client operation an error.  We just need to check both.  AFAICT `DeadlineExceeded` and `Canceled` are the only possible context errors.  Could probably just change this to `ctx.Err() != nil`, but since there is no way for the context to be cancelled here since the call to `cancel()` is a defer for this function, they are functionally equivalent.

Since all tests are green, I'd like to avoid non-functional changes if we can.  This ok?

Reply:
If the context is cancelled/timed out, ContainerList should return an error. Though, if it isn't, why do you want to abandon the result in this case?

Reply:
fair enough. updated.

Reply:
```go
func clientErr(ctx context.Context, err error) error {
	if ctx.Err() != nil {
		return ctx.Err()
	}
	return err
}
```
@sjenning You move an error handling to this function, it checks `ctx.Err()` first even if `err == nil` in the function arguments. I do not see any changes in the old logic.

Reply:
@sjenning ping